### PR TITLE
CAS Jetty Container Authentication Support

### PR DESCRIPTION
**Pull request by serac**
This is intended to satisfy #152. I've done a fair bit of testing with both CASv2 and SAML1.1 support and everything works as it should, including logout for servlet 3.0+ apps by calling `HttpServletRequest#logout()`. Notable implementation details:

1. Gateway is not supported, but I can't imagine a use case for it anyway.
2. Proxy support has not been tested but I don't think it would work without further effort.

In any case the feature set is similar to what we have for Tomcat. Hopefully the Jetty security components will prove more stable over time than Tomcat (which change every major version).

### CODE REVIEW

**Comment by serac** [comment]
@battags could you please let me know when you think you'll have time to review?

**Comment by battags** [comment]
Sorry -- hopefully this weekend!!  Its on my TODO list.

**Comment by serac** [comment]
No worries, thanks for the update.

**Comment by battags** [comment]
FYI -- I'm just starting to look at this (so comments may randomly trickle in)

File: cas-client-core/src/main/java/org/jasig/cas/client/util/ReflectUtils.java [hunk 1]
```diff
@@ -148,4 +149,35 @@ public static void setProperty(final String propertyName, final Object value, fi
...
+            } catch (NoSuchFieldException e) {
+                clazz = clazz.getSuperclass();
+            }
+        } while (field == null && clazz != null);
+        if (field == null) {
+            throw new IllegalArgumentException(fieldName + " does not exist on " + target);
+        }
+        try {
+            if (!field.isAccessible()) {
+                field.setAccessible(true);
```
Comment (battags) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/ReflectUtils.java]: does this need to be set back to false after the get(...) ? (I can't remember)

File: cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthentication.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * Creates a new instance.
+     *
+     * @param authenticator The authenticator that produced this authentication.
+     * @param ticket The CAS ticket that was successfully validated to permit authentication.
+     * @param assertion The CAS assertion produced from successful ticket validation.
+     */
+    public CasAuthentication(final CasAuthenticator authenticator, final String ticket, final Assertion assertion) {
+        super(authenticator.getAuthMethod(), new CasUserIdentity(assertion, authenticator.getRoleAttribute()));
+        assert ticket != null : "Ticket cannot be null";
```
Comment (battags) [inline cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthentication.java]: can we switch to the CommonUtils asserts?

File: cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    private final Logger logger = LoggerFactory.getLogger(CasAuthenticator.class);
+
+    /** Map of tickets to sessions. */
+    private final ConcurrentMap<String, WeakReference<HttpSession>> sessionMap =
+            new ConcurrentHashMap<String, WeakReference<HttpSession>>();
+
+    /** CAS ticket validator component. */
+    private TicketValidator ticketValidator;
+
+    /** Space-delimited ist of server names. */
```
Comment (battags) [inline cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java]: list

File: cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+                    casServerLoginUrl, protocol.getServiceParameterName(), serviceUrl(request, response), renew, false);
+            logger.debug("Redirecting to {}", redirectUrl);
+            response.sendRedirect(redirectUrl);
+        } catch (IOException e) {
+            logger.debug("Redirect to CAS failed with error: {}", e);
+            throw new ServerAuthException("Redirect to CAS failed", e);
+        }
+    }
+
+    private String extractTicket(final HttpServletRequest request) {
```
Comment (battags) [inline cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java]: should this use the stored protocol object?

File: cas-client-core/src/main/java/org/jasig/cas/client/util/ReflectUtils.java (hunk 1 above)
Comment (serac) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/ReflectUtils.java]: No. My reading of the Javadocs suggest it's just for the reference, https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/AccessibleObject.html#setAccessible(java.lang.reflect.AccessibleObject[],%20boolean); moreover, I've never seen it done in practice.

File: cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthentication.java (hunk 2 above)
Comment (serac) [inline cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthentication.java]: For consistency, sure, but I think assert is better in every way. We might consider a separate issue to move to the keyword across the board.

File: cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java (hunk 3 above)
Comment (serac) [inline cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java]: Nice catch. I had to read several times to see what you were pointing out, then finally saw the typo.

File: cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java (hunk 4 above)
Comment (serac) [inline cas-client-integration-jetty/src/main/java/org/jasig/cas/client/jetty/CasAuthenticator.java]: Good suggestion, don't even need that method now using instance field `protocol`. Fixed in c1c691d.

**Comment by battags** [comment]
I think most of my comments are pretty minor -- I think the only one that rises above minor is the comment about Protocol.

**Comment by serac** [comment]
I pushed commits to address all your feedback. Hopefully this is good to go.

**Comment by serac** [comment]
I've responded to all feedback and merged latest master changes onto the pull branch, so we should be good to merge this.

**Comment by battags** [comment]
+1

**Pull request merged**